parser/prism/Translator.cc

```diff
@@ -462,6 +462,21 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node) {
             auto params = translate(reinterpret_cast<pm_node *>(defNode->parameters));
             auto body = translate(defNode->body);
 
+            if (defNode->body != nullptr && PM_NODE_TYPE_P(defNode->body, PM_BEGIN_NODE)) {
+                // If the body is a PM_BEGIN_NODE instead of a PM_STATEMENTS_NODE, it means the method definition
+                // doesn't have an explicit begin block.
+                // If that's the case, we need to check if the body contains an ensure or rescue clause, and if so,
+                // we need to elevate that node to the top level of the method definition, without the Kwbegin node to
+                // match the behavior of Sorbet's legacy parser.
+                auto kwbeginNode = dynamic_cast<parser::Kwbegin *>(body.get());
+
+                if (kwbeginNode != nullptr && kwbeginNode->stmts[0] != nullptr &&
+                    (dynamic_cast<parser::Rescue *>(kwbeginNode->stmts[0].get()) != nullptr ||
+                     dynamic_cast<parser::Ensure *>(kwbeginNode->stmts[0].get()) != nullptr)) {
+                    body = move(kwbeginNode->stmts[0]);
+                }
```

Comment:
This looks similar to the `inlineIfSingle` logic of `Translator::translateStatements()`. Is it possible to re-use that here somehow?

Reply:
I can't think of any case that it'd have multiple `Ensure` or `Rescue` or other statements. 

If both `Ensure` and `Rescue` are present, `Rescue` will be wrapped inside `Ensure` with the actual method body:

```
       Ensure {
          body = Rescue {
            body = Begin { # Actual method body
              stmts = [
                String {
                  val = <U string1>
                }
                String {
                  val = <U string2>
                }
              ]
            }
            rescue = [
              Resbody {
                exception = NULL
                var = NULL
                body = String {
                  val = <U rescued rescue>
                }
              }
            ]
            else_ = String {
              val = <U rescued else>
            }
          }
          ensure = String {
            val = <U ensure>
          }
        }
```

But I can add an unreachable statement to help us debug future issues in case there are exceptions.